llvm/test/CodeGen/AArch64/vector-insert-dag-combines.ll

```diff
@@ -0,0 +1,310 @@
+; CHECK:       t11: nxv8i8 = extract_subvector t2, Constant:i64<0>
+; CHECK:     t7: nxv8i16 = any_extend t11
+; CHECK:   t9: ch,glue = CopyToReg t0, Register:nxv8i16 $z0, t7
+; CHECK:   t10: ch = AArch64ISD::RET_GLUE t9, Register:nxv8i16 $z0, t9:1
+
+define <vscale x 8 x i8> @extract_small_scalable_from_big_scalable_1(<vscale x 16 x i8> %a) #0 {
+  %extract = call <4 x i8> @llvm.vector.extract(<vscale x 16 x i8> %a, i64 0)
+  %insert = call <vscale x 8 x i8> @llvm.vector.insert(<vscale x 8 x i8> undef, <4 x i8> %extract, i64 0)
+  ret <vscale x 8 x i8> %insert
+}
+
+; CHECK: Initial selection DAG: %bb.0 'extract_small_scalable_from_big_scalable_2:'
+; CHECK: SelectionDAG has 11 nodes:
+; CHECK:   t0: ch,glue = EntryToken
+; CHECK:           t2: nxv16i8,ch = CopyFromReg t0, Register:nxv16i8 %0
+; CHECK:         t4: nxv4i8 = extract_subvector t2, Constant:i64<0>
+; CHECK:       t6: nxv8i8 = insert_subvector undef:nxv8i8, t4, Constant:i64<0>
+; CHECK:     t7: nxv8i16 = any_extend t6
+; CHECK:   t9: ch,glue = CopyToReg t0, Register:nxv8i16 $z0, t7
+; CHECK:   t10: ch = AArch64ISD::RET_GLUE t9, Register:nxv8i16 $z0, t9:1
+
+; CHECK: Optimized lowered selection DAG: %bb.0 'extract_small_scalable_from_big_scalable_2:'
+; CHECK: SelectionDAG has 9 nodes:
+; CHECK:   t0: ch,glue = EntryToken
+; CHECK:         t2: nxv16i8,ch = CopyFromReg t0, Register:nxv16i8 %0
+; CHECK:       t11: nxv8i8 = extract_subvector t2, Constant:i64<0>
+; CHECK:     t7: nxv8i16 = any_extend t11
+; CHECK:   t9: ch,glue = CopyToReg t0, Register:nxv8i16 $z0, t7
+; CHECK:   t10: ch = AArch64ISD::RET_GLUE t9, Register:nxv8i16 $z0, t9:1
+
+define <vscale x 8 x i8> @extract_small_scalable_from_big_scalable_2(<vscale x 16 x i8> %a) #0 {
+  %extract = call <vscale x 4 x i8> @llvm.vector.extract(<vscale x 16 x i8> %a, i64 0)
+  %insert = call <vscale x 8 x i8> @llvm.vector.insert(<vscale x 8 x i8> undef, <vscale x 4 x i8> %extract, i64 0)
+  ret <vscale x 8 x i8> %insert
+}
+
+; CHECK: Initial selection DAG: %bb.0 'extract_fixed_from_scalable:'
+; CHECK: SelectionDAG has 10 nodes:
+; CHECK:   t0: ch,glue = EntryToken
+; CHECK:         t2: nxv16i8,ch = CopyFromReg t0, Register:nxv16i8 %0
+; CHECK:       t4: v4i8 = extract_subvector t2, Constant:i64<0>
+; CHECK:     t6: v16i8 = insert_subvector undef:v16i8, t4, Constant:i64<0>
+; CHECK:   t8: ch,glue = CopyToReg t0, Register:v16i8 $q0, t6
+; CHECK:   t9: ch = AArch64ISD::RET_GLUE t8, Register:v16i8 $q0, t8:1
+
+; CHECK: Optimized lowered selection DAG: %bb.0 'extract_fixed_from_scalable:'
+; CHECK: SelectionDAG has 8 nodes:
+; CHECK:   t0: ch,glue = EntryToken
+; CHECK:       t2: nxv16i8,ch = CopyFromReg t0, Register:nxv16i8 %0
+; CHECK:     t10: v16i8 = extract_subvector t2, Constant:i64<0>
+; CHECK:   t8: ch,glue = CopyToReg t0, Register:v16i8 $q0, t10
+; CHECK:   t9: ch = AArch64ISD::RET_GLUE t8, Register:v16i8 $q0, t8:1
+
+; A variant of insert_small_scalable_into_big_fixed whose vector types prevent
```

Comment:
Is this actually testing what you want to test? It looks to me like this will immediately fail the test `VT.knownBitsGE(SrcVT)` because we cannot prove sizeof(<16 x i8>) >= sizeof(<vscale x 16 x i8>)? If you want to test the case where `VT.knownBitsGE(SrcVT)` is true *and* you're inserting a scalable vector into a fixed-length vector you would need a case where you can prove a fixed-width vector is always >= a scalable vector. That would require adding support to VT.knownBitsGE to take vscale_range into account I think?